### ISSUE TITLE
[0.17] Dependency missing in depcache despite having been installed by jspm itself

### ISSUE THREAD

**Issue by Rush** [jspm/jspm-cli#2197]
I'm running 0.17.0-beta.32.  The gyst of the problem is that jspm does not "see" installed dependencies of package `angular-ui-router` as shown on this generated depcache. Trying to run the app with this depcache will yield `angular-ui-router` being loaded without `angular` and breaking.

```
    "depCache": {
      "test.js": [
        "angular-ui-router"
      ]
```

Please see list of commands 

```
rush@rushpod: ~/c/jspm-bug 318ms » jspm init

warn Running jspm globally, it is advisable to locally install jspm via npm install jspm --save-dev.

Package.json file does not exist, create it? [Yes]: Yes
Init mode (Quick, Standard, Custom) [Quick]: Quick
Local package name (recommended, optional): app
package.json directories.baseURL: .
package.json configFiles folder [./]: ./
Use package.json configFiles.jspm:dev? [No]: No
SystemJS.config browser baseURL (optional): /
SystemJS.config Node local project path [src/]: src/
SystemJS.config local package main [app.js]: app.js
SystemJS.config transpiler (Babel, Traceur, TypeScript, None) [babel]: babel
     Updating registry cache...

ok   Installed dev dependency plugin-babel to npm:systemjs-plugin-babel@^0.0.17 (0.0.17)
     Install tree has no forks.
     
ok   jspm init completed. For the standard project layout see http://jspm.io/0.17-beta-guide/creating-a-project.html.
     To rerun these init prompts at any time use jspm init.
     
ok   package.json at package.json
     Config at jspm.config.js
     Looking up loader files...
       system.js
       system.js.map
       system-polyfills.js
       system-polyfills.js.map
       system.src.js
       system-polyfills.src.js
     
     Using loader versions:
       systemjs@0.19.41
ok   Loader files downloaded successfully
rush@rushpod: ~/c/jspm-bug 18.77s » 
rush@rushpod: ~/c/jspm-bug » jspm install angular-ui-router

warn Running jspm globally, it is advisable to locally install jspm via npm install jspm --save-dev.
     Updating registry cache...
     Downloading github:angular-ui/angular-ui-router-bower@0.3.2

warn Unable to use the GitHub API to speed up dependency downloads due to lack of authorization
     To resolve use jspm registry config github to configure the credentials, or update them in your ~/.netrc file.
ok   Installed github:angular/bower-angular (1.6.0)
ok   Installed angular-ui-router to github:angular-ui/angular-ui-router-bower@^0.3.2 (0.3.2)
     Install tree has no forks.

ok   Install complete.
rush@rushpod: ~/c/jspm-bug 9.33s » ls
jspm.config.js  jspm_packages/  package.json
rush@rushpod: ~/c/jspm-bug » echo 'import foo from "angular-ui-router"' > test.js
rush@rushpod: ~/c/jspm-bug » jspm depcache test.js
     Injecting the traced dependency tree for test.js...
     
       github:angular-ui/angular-ui-router-bower@0.3.2/release/angular-ui-router.js
       test.js
     
ok   Dependency tree injected
rush@rushpod: ~/c/jspm-bug 1.45s » cat jspm.config.js 
SystemJS.config({
  paths: {
    "npm:": "jspm_packages/npm/",
    "github:": "jspm_packages/github/",
    "app/": "src/"
  },
  browserConfig: {
    "baseURL": "/",
    "depCache": {
      "test.js": [
        "angular-ui-router"
      ]
    }
  },
  devConfig: {
    "map": {
      "plugin-babel": "npm:systemjs-plugin-babel@0.0.17"
    }
  },
  transpiler: "plugin-babel",
  packages: {
    "app": {
      "main": "app.js",
      "meta": {
        "*.js": {
          "loader": "plugin-babel"
        }
      }
    }
  }
});

SystemJS.config({
  packageConfigPaths: [
    "npm:@*/*.json",
    "npm:*.json",
    "github:*/*.json"
  ],
  map: {
    "angular-ui-router": "github:angular-ui/angular-ui-router-bower@0.3.2"
  },
  packages: {
    "github:angular-ui/angular-ui-router-bower@0.3.2": {
      "map": {
        "angular": "github:angular/bower-angular@1.6.0"
      }
    }
  }
});
```

I also uploaded above showcase here: http://x.rushbase.net/2623cd8bcaef14b2e4c4e7e48cb96cfb0d1d5d99/jspm-bug.zip

**Comment by joeldenning** [comment]
@Rush thanks for the example in the zip file - it helped me diagnose what is going on. It looks like the problem is that version 0.3.2 of angular-ui-router does not have an override in the jspm registry. Version 0.3.1 has an override (like you pointed out in gitter), but not 0.3.2 which was released last month.

I've created https://github.com/jspm/registry/pull/1014 to fix this. While you're waiting on that to be merged/published, you can achieve the same thing by putting the following into your jspm.config.js:

```js
SystemJS.config({
  packageConfigPaths: [
    "npm:@*/*.json",
    "npm:*.json",
    "github:*/*.json"
  ],
  map: {
    "angular-ui-router": "github:angular-ui/angular-ui-router-bower@0.3.2"
  },
  packages: {
    "github:angular-ui/angular-ui-router-bower@0.3.2": {
      "map": {
        "angular": "github:angular/bower-angular@1.6.0"
      },
      "meta": {
        "*": {
          "globals": {
            "angular": "angular"
          },
          "deps": [
            "angular"
          ]
        }
      }
    }
  }
});
```

The above code is the same as what you have in the zip file, except for the addition of the `meta` configuration of angular-ui-router. The reason that the meta configuration is necessary is that angular-ui-router references `window.angular` which is something that systemjs needs to know about when loading the angular-ui-router module.

In my testing, depcache did not have anything to do with the issue -- it didn't work regardless of depcache until I added the meta configuration.

**Comment by Rush** [comment]
Sadly it doesn't seem to work. Also, if I repeat the exact same steps and install `angular-ui-router@0.3.1` which should have an override already the depcache is still silent about angular:

```
    "depCache": {
      "test.js": [
        "angular-ui-router"
      ]
    }
```

You're right that depcache is not the root issue here. The root issue seems to be that jspm does not see `angular` as the dependency of `angular-ui-router`. Well it does see it at installation time, but not at runtime/"dependency injection".

**Comment by Rush** [comment]
Raw steps to reproduce the problem:
```
jspm init # use all default
jspm install angular-ui-router@0.3.1
echo 'import foo from "angular-ui-router"' > test.js
jspm depcache test.js
cat jspm.config.js # look for depcache entries
```

**Comment by joeldenning** [comment]
@Rush These steps work for me (and illustrate that my package override for ui-router@0.3.2 works:

```bash
jspm init
jspm install angular-ui-router@0.3.2 -o "{meta: {'*': {globals: {angular: 'angular'}, deps: ['angular']}}}"
echo 'import foo from "angular-ui-router"' > test.js
jspm depcache test.js
```

**Comment by joeldenning** [comment]
As for the angular-ui-router@0.3.1, I was able to confirm the behavior that you were seeing where it wasn't working. The reason is that jspm@0.17 upgrades [shim](https://github.com/jspm/registry/blob/master/package-overrides/github/angular-ui/angular-ui-router-bower%400.3.1.json#L3) to be a fancier [meta configuration](https://github.com/jspm/jspm-cli/commit/3944cb3bf346a0f47b107937e2de5683f96c944b#diff-8eaf344b0d37588331d22edb952ddd05R320), but that there is a bug.

The bug resulted in the override to be this:
```json
    "overrides": {
      "github:angular-ui/angular-ui-router-bower@0.3.1": {
        "dependencies": {
          "angular": "*"
        },
        "meta": {
          "release/angular-ui-router.js.js": {
            "deps": [
              "angular"
            ],
            "format": "global"
          }
        }
      },
      "github:angular/bower-angular@1.6.0": {
        "main": "angular",
        "meta": {
          "angular.js": {
            "exports": "angular",
            "format": "global"
          }
        }
      }
    }
```

Notice that `release/angular-ui-router.js.js` has `.js` twice, which is the bug. Once that is corrected, everything works correctly.

@Rush for your case, if you want to use angular-ui-router@0.3.2, run this `jspm install angular-ui-router@0.3.2 -o "{meta: {'*': {globals: {angular: 'angular'}, deps: ['angular']}}}"`. If you want to use 0.3.1, run this: `jspm install angular-ui-router@0.3.1 -o "{meta: {'*': {globals: {angular: 'angular'}, deps: ['angular']}}}"`

**Comment by Rush** [comment]
@joeldenning do you suggest closing this issue or keeping it open (perhaps with changing the title) ?

**Comment by joeldenning** [comment]
I would say close it -- the issue reported has been resolved. I still plan on pull requesting the [jspm registry for ui-router@0.3.1](https://github.com/jspm/registry/blob/master/package-overrides/github/angular-ui/angular-ui-router-bower%400.3.1.json), but will do that even after this issue is closed.

**Comment by aluanhaddad** [comment]
@joeldenning What needs to be fixed in that override? Is it just changing it to match the one for `for ui-router@0.3.2`? Are there any 0.3.1 specific needs that need to be addressed?

**Comment by joeldenning** [comment]
@aluanhaddad the `.js` extension found [here](https://github.com/jspm/registry/blob/master/package-overrides/github/angular-ui/angular-ui-router-bower%400.3.1.json#L4) should not be there. See the discussion in https://github.com/jspm/jspm-cli/pull/2200 where guy bedford explains this

**Comment by joeldenning** [comment]
I created the pull request to fix angular-ui-router@0.3.1 in jspm@0.17. See https://github.com/jspm/registry/pull/1019

**Comment by aluanhaddad** [comment]
Fixed by #1019

**Comment by Rush** [comment]
Thank you guys! You're awesome.
